demo.ipynb

```diff
@@ -0,0 +1,1193 @@
+   "metadata": {},
+   "source": [
+    "At this point, the privacy budget of the context, configured at the start with `epsilon` and `weights`, is exhausted: attempting to make another release will result in an error."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "---\n",
+    "\n",
+    "# Other exports\n",
+    "\n",
+    "Below is a proposal for what the other export formats (text and CSV) would look like. We first make a data structure with everything we need, and then use generic methods to serialize that structure."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "{'inputs': {'csv_path': '/tmp/demo.csv',\n",
+       "  'contributions': 10,\n",
+       "  'epsilon': 2,\n",
+       "  'weights': [4, 4, 1, 1],\n",
+       "  'max_possible_rows': 1000000,\n",
+       "  'delta': 1e-07,\n",
+       "  'grade': {'min': 50, 'max': 100, 'bins_count': 10},\n",
+       "  'class_year': {'min': 1, 'max': 4, 'bins_count': 4}},\n",
+       " 'outputs': {'grade': {'mean': 81.65544332210999,\n",
+       "   'histogram': {'(55, 60]': 22,\n",
+       "    '(60, 65]': 0,\n",
+       "    '(65, 70]': 60,\n",
+       "    '(70, 75]': 105,\n",
+       "    '(75, 80]': 196,\n",
+       "    '(80, 85]': 199,\n",
+       "    '(85, 90]': 205,\n",
+       "    '(90, 95]': 74,\n",
+       "    '(95, inf]': 41}},\n",
+       "  'class_year': {'mean': 1.7699214365881033,\n",
+       "   'histogram': {'(-inf, 1]': 427,\n",
+       "    '(1, 2]': 304,\n",
+       "    '(2, 3]': 138,\n",
+       "    '(3, inf]': 61}}}}"
+      ]
+     },
+     "execution_count": 21,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "release = {\n",
+    "    'inputs': {\n",
+    "        'csv_path': csv_path,\n",
+    "        'contributions': contributions,\n",
+    "        'epsilon': epsilon,\n",
+    "        'weights': weights,\n",
+    "        'max_possible_rows': max_possible_rows,\n",
+    "        'delta': delta,\n",
+    "        'grade': {\n",
+    "            'min': grade_min,\n",
+    "            'max': grade_max,\n",
+    "            'bins_count': grade_bins_count,\n",
+    "        },\n",
+    "        'class_year': {\n",
+    "            'min': class_year_min,\n",
+    "            'max': class_year_max,\n",
+    "            'bins_count': class_year_bins_count,\n",
+    "        }    \n",
+    "    },\n",
+    "    'outputs': {\n",
+    "        'grade': {\n",
+    "            'mean': grade_mean.item(),\n",
+    "            'histogram': {v['grade_bin']: v['len'] for v in grade_histogram.to_dicts()}\n",
+    "        },\n",
+    "        'class_year': {\n",
+    "            'mean': class_year_mean.item(),\n",
+    "            'histogram': {v['class_year_bin']: v['len'] for v in class_year_histogram.to_dicts()}\n",
+    "        },\n",
+    "    }\n",
+    "}\n",
```

Comment:
Hmm, while logger output could be seen as another medium for specifying queries, I don't think that format is very human-editable, so I'd want to avoid logger tabs. If the lower editability is not consistent with your vision for this json output, that's fine.

Reply:
Yeah, it would be good to spruce the package up and add to docs.opendp.org. It always seems to come up when people want to integrate with OpenDP.